### PR TITLE
Martin/ios 553 add the ability to send links

### PR DESCRIPTION
**Pull request by MartinJibber**

### CODE REVIEW

File: iOS/Managers/AttachmentsManager.swift [hunk 1]
```diff
@@ -58,36 +58,39 @@ class AttachmentsManager {
...
-                        try previewData?.write(to: previewURL, options: .atomic)
-                        let item = PhotoAttachment(url: url, previewUrl: previewURL, _data: data, info: info)
-                        continuation.resume(returning: .photo(photo: item, body: body))
-                    } catch  {
-                        logError(error)
-                        continuation.resume(throwing: error)
-                    }
-                default:
-                    continuation.resume(throwing: ClientError.message(detail: "Unknown asset type."))
+            guard let mediaType = info[.mediaType] as? String else {
```
Comment (MartinJibber) [inline iOS/Managers/AttachmentsManager.swift:65]: Just using a guard here. No functional change.

File: Shared/Extensions/String Extensions/String+DataTypes.swift [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    /// Returns all the URLs found in the string.
+    func getURLs() -> [URL] {
+        let results = self.getDataTypes(with: NSTextCheckingResult.CheckingType.link.rawValue)
+        let urls = results.compactMap { textCheckingResult in
+            return textCheckingResult.url
+        }
+
+        return urls
+    }
```
Comment (MartinJibber) [inline Shared/Extensions/String Extensions/String+DataTypes.swift:48]: Useful functions for dealing with urls in strings.

File: Shared/Models/Message Model/MessageKind.swift [hunk 3]
```diff
@@ -60,19 +61,12 @@ enum MessageKind {
...
     }
-
-    var displayable: ImageDisplayable? {
-        switch self {
-        case .photo(photo: let photo, _):
-            return photo.image
-        default:
-            return nil
-        }
-    }
```
Comment (MartinJibber) [inline Shared/Models/Message Model/MessageKind.swift:75]: Unused.

File: Shared/Models/Message Model/MessageKind.swift [hunk 4]
```diff
@@ -94,14 +88,42 @@ extension MessageKind: Equatable {
             return lhsAudio == rhsAudio
         case (.contact(let lhsContact), .contact(let rhsContact)):
             return lhsContact == rhsContact
-        case (.link(let lhsLink), .link(let rhsLink)):
-            return lhsLink == rhsLink
+        case (.link(let lhsURL, let lhsStringURL), .link(let rhsURL, let rhsStringURL)):
+            return lhsURL == rhsURL && lhsStringURL == rhsStringURL
```
Comment (MartinJibber) [inline Shared/Models/Message Model/MessageKind.swift:92]: Distinguishing between the message text, and the cleaned up/sanitized url object.

File: iOS/Flows/Conversation/Attachments/AttachmentsCoordinator.swift [hunk 5]
```diff
@@ -19,12 +19,12 @@ enum AttachmentOption {
 
 class AttachmentsCoordinator: PresentableCoordinator<AttachmentOption> {
     
-    lazy var attachementsVC = AttachmentsViewController()
+    lazy var attachmentsVC = AttachmentsViewController()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Attachments/AttachmentsCoordinator.swift:22]: Spelling

File: iOS/Flows/Conversation/Conversation Header/ConversationHeaderViewController.swift [hunk 6]
```diff
@@ -71,7 +71,7 @@ class ConversationHeaderViewController: ViewController, ActiveConversationable {
         
         self.jibImageView.squaredSize = 44
         self.jibImageView.pin(.right, offset: .custom(6))
-        self.jibImageView.pin(.top, offset: .custom(16))
+        self.jibImageView.centerOnY()
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Conversation Header/ConversationHeaderViewController.swift:74]: Moving up the jib view so all of the header Subviews fit inside of it.

File: iOS/Flows/Conversation/Input/SwipeableInputAccessoryViewController.swift [hunk 7]
```diff
@@ -253,8 +252,14 @@ class SwipeableInputAccessoryViewController: UIInputViewController {
...
-        case .text(_):
-            self.currentMessageKind = .text(text)
+        case .text, .link:
+            // If there's URL in the text, then this becomes a link message.
+            if text.isSingleLink, let url = text.getURLs().first {
+                self.currentMessageKind = .link(url: url, stringURL: text)
+            } else {
+                // No URL means we're just sending plain text.
+                self.currentMessageKind = .text(text)
+            }
```
Comment (MartinJibber) [inline iOS/Flows/Conversation/Input/SwipeableInputAccessoryViewController.swift:262]: Detecting the text is a link.

File: iOS/_Stream Temp/ChatMessage+Extensions.swift [hunk 8]
```diff
@@ -88,29 +88,31 @@ extension Message: Messageable {
...
-            case AttachmentType.audio:
-                break
-            case AttachmentType.file:
-                break
-            case AttachmentType.giphy:
-                break
-            case AttachmentType.linkPreview:
-                break
-            default:
+        if let imageAttachment = self.imageAttachments.first {
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessage+Extensions.swift:91]: Prioritize images first.

File: iOS/_Stream Temp/ChatMessage+Extensions.swift [hunk 9]
```diff
@@ -88,29 +88,31 @@ extension Message: Messageable {
...
+            let attachment = PhotoAttachment(url: imageAttachment.imageURL,
+                                             previewUrl: imageAttachment.imagePreviewURL,
+                                             _data: nil,
+                                             info: nil)
+            return .photo(photo: attachment, body: self.text)
+        } else if self.text.isSingleLink,
+                  self.linkAttachments.count == 1,
+                  let linkAttachment = self.linkAttachments.first {
+
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessage+Extensions.swift:100]: Make sure it's just a link and no other text.

File: iOS/_Stream Temp/ChatMessage+Extensions.swift [hunk 10]
```diff
@@ -88,29 +88,31 @@ extension Message: Messageable {
...
+            
+            guard var urlComponents = URLComponents(url: linkAttachment.originalURL,
+                                                    resolvingAgainstBaseURL: false) else {
                 return .text(self.text)
             }
+
+            // Make sure the url has a scheme.
+            if urlComponents.scheme.isNil {
+                urlComponents.scheme = "https"
+            }
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ChatMessage+Extensions.swift:109]: Fixes issues with unfinished urls like "google.com"

File: iOS/_Stream Temp/ConversationController+Extensions.swift [hunk 11]
```diff
@@ -142,39 +142,31 @@ extension ConversationController {
...
-            break
-        case .contact:
-            break
-        case .link:
-            break
+            messageBody = body
+        case .link(_, let stringURL):
+            // The link URL is automatically detected by stream and added as an attachment.
+            // Removing extra whitespace and make links lower case.
+            messageBody = stringURL.trimWhitespace().lowercased()
```
Comment (MartinJibber) [inline iOS/_Stream Temp/ConversationController+Extensions.swift:162]: See comment

**Review by Benjamindodgson — APPROVED** [review]

**Pull request merged**